### PR TITLE
AI-3204: surface Snowflake job_id to MCP clients via progress notification

### PR DESCRIPTION
**Pull request by Matovidlo**
## Description

**Linear**: [AI-3204](https://linear.app/keboola/issue/AI-3204/cancel-query-data-queries-when-chat-is-stopped) — follow-up to #526.

### Change Type

- [ ] Major (breaking changes, significant new features)
- [ ] Minor (new features, enhancements, backward compatible)
- [x] Patch (bug fixes, small improvements, no new features)

### Summary

Enables MCP clients to cancel a long-running `query_data` call **directly against Query Service**, bypassing the originating MCP replica. This closes the cross-replica gap that PR #526's in-memory cancellation registry leaves open.

**How it works**

1. **Server surfaces the backend job handle on the response stream.** `_SnowflakeWorkspace.execute_query` fires a new `on_job_submitted` callback the moment Query Service returns a `queryJobId`. `query_data` turns that into an MCP `notifications/progress` on POST #1's SSE stream, with `params._meta` carrying:
   - `keboola.queryJobId`
   - `keboola.backend` (e.g. `"snowflake"`)
   - `keboola.cancellationUrl` (absolute URL like `https://query.<stack>.keboola.com/api/v1/queries/<jobId>/cancel`)

2. **Clients call QS directly to cancel** — using the same auth header they already use to talk to the MCP server. `RawKeboolaClient` auto-detects the token shape (`clients/base.py:38-42`): `Authorization: Bearer <oauth-token>` for Claude Code / claude.ai, `X-StorageAPI-Token: <pat>` for Kai / in-platform agent. No special-case auth path on either side.

3. **Original `query_data` self-cleans.** The Snowflake poll loop already exits when the job transitions to `canceled` (`workspace.py:260-264`). POST #1's response stream finalizes with an error result — no `asyncio.CancelledError` plumbing, no shared store, no sticky LB routing required.

### What this fixes

PR #526's `cancellation.py` registry handles the same-replica case correctly, but in the production multi-replica deployment behind `mcp.keboola.com` the cancel `POST /mcp` is an independent HTTP request that the LB hashes separately from the original `tools/call`. When it lands on a different replica, the registry there is empty, the cancel is dropped, and the query keeps running. This PR removes the need for cancel notifications to find the right replica at all: clients hold the handle and act on it directly.

### Backends

Both Snowflake and BigQuery now execute through Query Service (`_Workspace.execute_query` →
`QueryServiceClient.submit_job`), so both receive a real `queryJobId` and the `on_job_submitted`
callback fires for either backend. `build_cancel_url` is backend-agnostic — it points at the
Query Service cancel endpoint `…/api/v1/queries/{jobId}/cancel`, which accepts the job handle
regardless of the underlying engine.

- **Snowflake**: full support. Callback fires after `submit_job`, cancellation URL surfaced.
- **BigQuery**: full support via the same shared Query Service path — callback fires, cancellation URL surfaced. (The job runs as a QS query job, so it is cancellable out-of-band exactly like Snowflake.)

### What's NOT in this PR

- **Client-side adoption** (Kai, Claude Code, claude.ai connector) — separate task per client. Stock clients that ignore unknown `_meta` keys are unaffected — they simply don't get out-of-band cancellation.

> Note: the original `cancellation.py` in-memory registry from #526 is **not present on this branch**;
> this PR's out-of-band contract (clients hold the handle and POST the cancel URL directly) supersedes it.

### Notes for client implementers

The notification follows the standard MCP `notifications/progress` shape. Spec-compliant clients that don't recognize the Keboola-specific `_meta` keys will simply ignore them and surface "Submitted to snowflake" as progress text. Clients that *do* read them should:

1. Persist `keboola.queryJobId` and `keboola.cancellationUrl` for the duration of the call.
2. On user-initiated stop, POST `{}` (or a JSON body with an optional `reason`) to `cancellationUrl` with the same auth header used for MCP.
3. The original `tools/call` SSE stream will finalize with an error result within one poll tick (~1-2s) of the Snowflake job actually canceling — no additional handling needed.

## Testing

- [ ] Tested with Cursor AI desktop (`Streamable-HTTP` transports)

### Optional testing
- [ ] Tested with Cursor AI desktop (all transports)
- [ ] Tested with claude.ai web and `canary-orion` MCP (`Streamable-HTTP`)
- [ ] Tested with In Platform Agent on `canary-orion`
- [ ] Tested with RO chat on `canary-orion`

**Unit tests added — full `tox` green (pytest + black + isort + flake8 + check-tools-docs):**

- `tests/test_workspace.py`:
  - `on_job_submitted` fires exactly once with the full `JobSubmittedInfo` carrying the cancel URL.
  - Callback omitted → `build_cancel_url` is not called (no wasted work).
  - Callback exception is suppressed → the query still completes.
  - `build_cancel_url` produces the expected absolute URL shape.
  - `WorkspaceManager.execute_query` forwards the callback to the active workspace.

- `tests/tools/test_sql.py`:
  - When `progressToken` is provided, `query_data` emits `notifications/progress` with `_meta.keboola.queryJobId` (+ backend, cancellationUrl).
  - When `progressToken` is absent, no notification is sent (MCP spec compliance).
  - On-wire JSON uses the `_meta` alias, so spec-compliant clients can parse it per spec.

**End-to-end verification recommended before merging:** start a deliberately heavy query against a real Snowflake workspace, capture the progress notification from the SSE stream, POST to the surfaced `cancellationUrl` with the user's token, and confirm the Snowflake query history shows `CANCELED` and the MCP response stream finalizes with an error result.

## Checklist

- [x] Self-review completed
- [x] Unit tests added/updated (if applicable)
- [ ] Integration tests added/updated (if applicable)
- [x] Project version bumped according to the change type
- [ ] Documentation updated (if applicable)

### CODE REVIEW

**Comment by Matovidlo** [comment]
@claude review

File: src/keboola_mcp_server/workspace.py [hunk 1]
```diff
@@ -257,6 +293,15 @@ async def execute_query(
...
         job_id = await self._qsclient.submit_job(statements=[sql_query], workspace_id=str(self.id))
+        if on_job_submitted is not None:
+            info = JobSubmittedInfo(
+                job_id=job_id,
+                cancellation_url=self._qsclient.build_cancel_url(job_id),
+                backend='snowflake',
+            )
+            # Best-effort: a failed progress notification must not kill the running query.
+            with contextlib.suppress(Exception):
+                await on_job_submitted(info)
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/workspace.py:249]: **Fixed** ✅ in commit ef5163e2d7146c7e086130f459dfa8e68c483ec6

Two things worth separating here:

**On the premise.** `asyncio.CancelledError` inherits from `BaseException` (not `Exception`) since Python 3.8, and our `requires-python = ">=3.10"`, so `contextlib.suppress(Exception)` does *not* swallow cancellation on any supported interpreter — the cancel propagates through and the tool task aborts as intended. This is documented in the Python 3.8 release notes ("asyncio.CancelledError now derives from BaseException instead of Exception") and is the load-bearing reason for the inheritance change.

**On the spirit.** Even though the premise is wrong for the supported range, making the intent explicit is cheap defensive documentation against a future refactor that might widen the catch to `BaseException`. So I replaced `contextlib.suppress(Exception)` with an explicit two-arm try/except:

```python
try:
    await on_job_submitted(info)
except asyncio.CancelledError:
    raise
except Exception as exc:
    LOG.warning(f'on_job_submitted callback raised for job_id={job_id}: {exc!r} — continuing')
```

The `except asyncio.CancelledError: raise` arm is functionally redundant on Python ≥3.10 but documents the invariant explicitly, and a comment in-code points at the BaseException inheritance rule so the next maintainer doesn't think the explicit arm is the load-bearing one. Also dropped the now-unused `contextlib` import.

File: src/keboola_mcp_server/workspace.py [hunk 2]
```diff
@@ -733,10 +787,29 @@ async def _get_workspace(self) -> _Workspace:
...
+        )
+
+    async def get_backend(self) -> str:
+        """Returns a short backend identifier (e.g. 'snowflake', 'bigquery') for the active workspace."""
+        workspace = await self._get_workspace()
+        if isinstance(workspace, _SnowflakeWorkspace):
+            return 'snowflake'
+        if isinstance(workspace, _BigQueryWorkspace):
+            return 'bigquery'
+        return type(workspace).__name__
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/workspace.py]: **Fixed** ✅ in commit ef5163e2d7146c7e086130f459dfa8e68c483ec6

Correct — added speculatively in the first cut on the assumption a downstream caller would want backend-name introspection, but the actual data flow surfaces the backend via `JobSubmittedInfo.backend` (set inside the workspace itself), so no caller ever materialised. Removed `WorkspaceManager.get_backend()` entirely; the public surface is smaller and there's no risk of `type(workspace).__name__` leaking an internal class name like `_SnowflakeWorkspace` to a consumer.

File: tests/test_workspace.py [hunk 3]
```diff
@@ -203,6 +233,112 @@ def _rebind(target_branch_id: str | None) -> Mock:
...
+    qs_mock.get_job_status.return_value = {
+        'status': 'completed',
+        'statements': [{'id': 'stmt-1'}],
+    }
+    qs_mock.get_job_results.return_value = {
+        'status': 'completed',
+        'message': 'ok',
+        'numberOfRows': 1,
+        'columns': [{'name': 'col'}],
+        'rows': [['v']],
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply tests/test_workspace.py]: **Fixed** ✅ in commit 84e4f41cf45beb0030e87b9bf2fa8e3a4988d2ee

Good catch. Production reads `results.get('data', [])` (workspace.py around the page-fetch loop), so the mock with `rows` was a silent regression hazard — the page-data loop would always default to `[]` and the test would never exercise the actual iteration. Renamed the key to `data` and added a comment in the helper noting that the mock shape is deliberately aligned with what `_SnowflakeWorkspace.execute_query()` reads, so a future rename will fail the test instead of being papered over by `.get(..., [])`.

File: src/keboola_mcp_server/tools/sql.py [hunk 4]
```diff
@@ -18,6 +24,56 @@
...
+    """
+    params = ProgressNotificationParams.model_validate(
+        {
+            'progressToken': progress_token,
+            'progress': 0,
+            'message': f'Submitted to {info.backend}',
+            '_meta': {
+                'keboola.queryJobId': info.job_id,
+                'keboola.backend': info.backend,
+                # cancellation_url is None on backends without an out-of-band cancel handle (e.g. BigQuery via SAPI);
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/sql.py]: **Fixed** ✅ in commit 84e4f41cf45beb0030e87b9bf2fa8e3a4988d2ee

Right — `_BigQueryWorkspace.execute_query()` takes the `on_job_submitted` parameter for interface parity only and never invokes it, so the "still emit" wording doesn't apply to BQ. Reworded the inline comment to be backend-agnostic — `cancellation_url may be None when a backend does not expose an out-of-band cancel endpoint; clients should treat the field as optional` — and dropped the BigQuery example. If/when a backend ever needs to emit a progress notification without a cancel handle, the comment now accurately describes the contract; until then it's a no-op fact that doesn't reference a misleading example.

File: src/keboola_mcp_server/workspace.py [hunk 5]
```diff
@@ -20,6 +20,27 @@
...
+
+    job_id: str
+    cancellation_url: str | None
+    backend: str
+
+
+JobSubmittedCallback = Callable[[JobSubmittedInfo], Awaitable[None]]
+"""Async callback invoked exactly once per execute_query, immediately after the backend
+returns a job handle. Callbacks are best-effort: any exception raised inside is suppressed
+so a failed progress notification cannot abort the underlying query."""
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/workspace.py]: **Fixed** ✅ in commit 273340ba60422d29a1fcc3362a143d7f4175e847

Correct — a triple-quoted string following an assignment is not a docstring, it's a standalone expression statement that gets evaluated and discarded. Python's docstring rule only attaches the first string literal inside a module/class/function body; outside of those positions it's a no-op (and the runtime cost, while tiny, is real). Converted to a `#` comment block above the `JobSubmittedCallback` assignment so the documentation lives at the discoverable position and there's no dangling expression.

File: src/keboola_mcp_server/tools/sql.py [hunk 6]
```diff
@@ -18,6 +24,56 @@
...
+                # `cancellation_url` may be None when a backend does not expose an out-of-band
+                # cancel endpoint; clients should treat the field as optional.
+                'keboola.cancellationUrl': info.cancellation_url,
+            },
+        }
+    )
+    notification = ProgressNotification(method='notifications/progress', params=params)
+    request_id = getattr(ctx, 'request_id', None)
+    await ctx.session.send_notification(ServerNotification(notification), related_request_id=request_id)
+    LOG.info(
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/sql.py:89]: **Fixed** ✅ in commit 2e777424fd7bb3406c068ddf553cdaa3bd380946

Good catch — this is exactly the failure mode the routing fix was supposed to prevent, and `getattr(ctx, 'request_id', None)` would re-introduce it through the back door if `request_id` is ever missing (transports where it isn't populated, edge cases in tests, etc.).

Guarded explicitly: if `request_id is None`, log a `WARNING` mentioning the `job_id` (so the operator can correlate against Snowflake) and return without calling `send_notification`. Sending with `related_request_id=None` would just re-create the GET_STREAM_KEY silent-drop bug.

The underlying query continues normally — surfacing the cancellation handle is best-effort, and a failure to emit must not abort the query. Out-of-band cancellation simply becomes unavailable for that call, with the warning making the degradation visible in the logs instead of swallowed.

Added a regression test `test_query_data_skips_progress_when_request_id_missing` that asserts: (a) when `ctx.request_id is None`, `session.send_notification` is NOT called; (b) the query still completes and returns a `QueryDataOutput`; (c) the warning log mentions both `job_id` and `request_id is None` so a future refactor can't silently regress to the unguarded form.

File: src/keboola_mcp_server/tools/sql.py [hunk 7]
```diff
@@ -95,7 +161,18 @@ async def query_data(
...
+        await _emit_job_submitted_progress(ctx, progress_token, info)
+
+    result = await workspace_manager.execute_query(
+        sql_query,
+        max_rows=MAX_ROWS,
+        max_chars=MAX_CHARS,
+        on_job_submitted=_on_job_submitted if progress_token is not None else None,
+    )
     LOG.info(' '.join(filter(None, [f'Query "{query_name}" executed successfully.', result.message])))
     if result.is_ok:
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/sql.py:184]: **Fixed** ✅ in commit b3e32f935cc680a7ee592f3b032ef1a98ac60f41

Good catch — the `LOG.info("... executed successfully")` line was unconditional, before the `result.is_ok` check, so failures and cancellations were logged as successes. Moved the success log inside the `if result.is_ok:` branch. The error branch now logs genuine failures at WARNING (`Query "..." failed.`) and the expected client-initiated cancel at INFO (`Query "..." was cancelled.`), so out-of-band STOPs don't pollute error dashboards.

File: src/keboola_mcp_server/tools/sql.py [hunk 8]
```diff
@@ -18,6 +24,66 @@
...
+    """Returns the progress token the client included in the original `tools/call`, or None.
+
+    Per MCP spec, a server may only send `notifications/progress` for a request when the client
+    explicitly provided a `progressToken` in that request's `_meta`. Tools that fall through here
+    without a token must stay silent — emitting unsolicited progress can confuse strict clients.
+    """
+    rc = ctx.request_context
+    if rc is None or rc.meta is None:
+        return None
+    return rc.meta.progressToken
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/sql.py]: **Fixed** ✅ in commit b3e32f935cc680a7ee592f3b032ef1a98ac60f41

Made the lookup defensive: `_client_progress_token` now returns `getattr(rc.meta, 'progressToken', None)` instead of `rc.meta.progressToken`. The typed `RequestParams.Meta` always carries the attribute (default `None`), but this guards the case where `_meta` is surfaced without `progressToken`, returning `None` and staying silent per spec rather than raising `AttributeError`.

File: src/keboola_mcp_server/workspace.py [hunk 9]
```diff
@@ -205,6 +234,23 @@ async def execute_query(
 
         ts_start = time.perf_counter()
         job_id = await self._qsclient.submit_job(statements=[sql_query], workspace_id=str(self.id))
+        if on_job_submitted is not None:
+            info = JobSubmittedInfo(
+                job_id=job_id,
+                cancellation_url=self._qsclient.build_cancel_url(job_id),
+                backend=self.get_sql_dialect().lower(),
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/workspace.py:241]: **Intentional** ℹ️

You're right that the callback fires for BigQuery too — and that is the intended behavior, not a leak. Both backends now run through `_Workspace.execute_query` → `QueryServiceClient.submit_job`, so BigQuery also gets a real `queryJobId`, and `build_cancel_url` points at the backend-agnostic QS cancel endpoint (`…/api/v1/queries/{jobId}/cancel`) which accepts the handle regardless of engine. So emitting for BigQuery is correct — those jobs are cancellable out-of-band exactly like Snowflake. The PR description was stale (it described the old #526-era SAPI synchronous-proxy path); I've updated the Backends section to state both backends are fully supported via the shared QS path.

File: src/keboola_mcp_server/tools/sql.py [hunk 10]
```diff
@@ -95,7 +161,18 @@ async def query_data(
     * Ensure valid filtering by checking actual data values first
     """
     workspace_manager = WorkspaceManager.from_state(ctx.session.state)
-    result = await workspace_manager.execute_query(sql_query, max_rows=MAX_ROWS, max_chars=MAX_CHARS)
+
+    progress_token = _client_progress_token(ctx)
+
+    async def _on_job_submitted(info: JobSubmittedInfo) -> None:
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply src/keboola_mcp_server/tools/sql.py:175]: **Already addressed** ✅ in commit b3e32f935cc680a7ee592f3b032ef1a98ac60f41

Correct — `cancellation.py` (the #526 in-memory registry) is not present on this branch, and `query_data` no longer registers in-flight tasks. The out-of-band contract in this PR (clients hold the job handle and POST the cancel URL directly) supersedes the registry rather than keeping it as a fallback. I've updated the PR description: removed the "kept as fallback" / "same-replica registry path" claims and added a note that the registry is not on this branch.
